Sources/Commands/SwiftPackageTool.swift

```diff
@@ -1207,8 +1207,7 @@ final class PluginDelegate: PluginInvocationDelegate {
     }
     
     private func doTest(subset: PluginInvocationTestSubset, parameters: PluginInvocationTestParameters) throws -> PluginInvocationTestResult {
-        // FIXME: To implement this we should factor out a lot of the code in SwiftTestTool.
-        throw StringError("Running tests from a plugin is not yet implemented")
+        return try SwiftTestTool.doTest(subset: subset, parameters: parameters, swiftTool: swiftTool)
```

Comment:
The implementation is in SwiftTestTool, where the existing support code is.  That code should really be moved down to one of the library modules where it can be used by both SwiftTestTool and the other code that needs it.  SwiftTestTool should just be a thin layer that defines command line actions that then call down into the library code that can run the tests without assumption about the context.

Reply:
agreed this seems like a bit of an abstraction leak. how difficult is it to refactor to make this better layered?

Reply:
It doesn't seem difficult, but will take a little time and introduce some risk to the existing test functionality.  I think the best would be to rewrite the testing functionality — it's a bit of a mess, with layers of changes.  Rewriting the testing seemed to me to be appropriate to do as a clean-up task, but I didn't want to block the ability of plugins to invoke tests on that.